File: yafu.py
#!/usr/bin/python3

# yafu - command line client for YaFU
# Copyright (C) 2016 Benjamin Abendroth
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
# 
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import argparse, requests, re, pickle
from time import time
from datetime import datetime, timedelta
from os.path import expanduser, isfile, basename, devnull
from urllib.parse import urlencode
from string import Template

DEFAULT_FORMAT = '''=== $basename $expired_str===
File: $file
Date: $date
Expires: $expire_date [$expires]
Type: $type
URL: $direct_url
Info: $info_url
Delete: $deletion_url
'''

RECORD_SHORTCUTS = {
   'b': 'basename',     'f': 'file',            'h': 'hidden',
   'i': 'info_url',     'd': 'deletion_url',    'u': 'direct_url',
   'D': 'date',         'e': 'expire_date',     'x': 'expired',
   'E': 'expires'
}

parser = argparse.ArgumentParser(prog='YaFU', description='Commandline client for YaFU')

# FILE OPTIONS
fileopts = parser.add_argument_group(title='File Options')
fileopts.add_argument('--db',    help='Database file', type=expanduser)
fileopts.add_argument('--no-db', help='Disable database', dest='db', action='store_const', const=devnull)

# ACTIONS
actions = parser.add_argument_group(title='Actions')
actions = actions.add_mutually_exclusive_group(required=True)
actions.add_argument('-d', '--delete',  help='Delete the given links', metavar='URL', nargs='+')
actions.add_argument('-u', '--upload',  help='Files to upload', metavar='FILE', nargs='+')
actions.add_argument('-l', '--list',    help='List the records in database', action='store_true')

# UPLOAD ACTION
ulcmd = parser.add_argument_group('Upload Options')
ulcmd.add_argument('-b', '--base-url',  help='Specify base-url')
ulcmd.add_argument('-e', '--email',     help='Specify email for file upload')
ulcmd.add_argument('-p', '--password',  help='Specify password for file upload')
ulcmd.add_argument('-c', '--comment',   help='Specify comment for file upload')
ulcmd.add_argument('-x', '--expires',   help='Specify expire time',
   choices=['30m', '1h', '6h', '1d', '3d', '1w', 'max'])

hide_group = ulcmd.add_mutually_exclusive_group()
hide_group.add_argument('--private', help='Hide file in public list [DEFAULT]', dest='hide', action='store_true')
hide_group.add_argument('--public',  help='Show file in public list', dest='hide', action='store_false')

# LIST ACTION
listcmd = parser.add_argument_group('List Options')
listcmd.add_argument('--show-expired',   help='List also expired records', action='store_true')
listcmd.add_argument('--date-format',    help='Specify date format')
listcmd.add_argument('-n', '--number',   help='Only list the last N records', type=int)
listcmd.add_argument('-f', '--format',   help='Specify output format', type=Template)

parser.set_defaults(
   db =           '~/.config/yafu.db', 
   base_url =     'http://pixelbanane.de/yafu',
   format =       DEFAULT_FORMAT,
   date_format =  '%Y-%m-%d %H:%M:%S',
   hide =         True,
   expires =      '1w'
)

class RecordExpander(dict):

   def __init__(self, record, args):
      self._record = record
      self._cache = {}
      self._args = args

   def __getitem__(self, var):

      if len(var) == 1:
         try:
            var = RECORD_SHORTCUTS[var]
         except:
            raise Exception('Record shortcut not found: ' + var)

      if var in self._record:
         return self._record[var]

      if var not in self._cache:
         self._cache[var] = self.getVar(var)

      return self._cache[var]

   def getVar(self, var):

      if var == 'expired':
         return (self._record['expires'] != 'max' and
                 time() > self._record['expire_ts'])

      if var == 'expired_str':
         return 'EXPIRED ' if self['expired'] else ''

      if var == 'type':
         return 'Private' if self._record['hide'] else 'Public'

      if var == 'date':
         return datetime.fromtimestamp(self._record['date_ts']).strftime(self._args.date_format)

      if var == 'expire_date':
         if self._record['expire_ts'] == 0:
            return 'NEVER'

         return datetime.fromtimestamp(self._record['expire_ts']).strftime(self._args.date_format)
         
      if var == 'direct_url':
         return '%s/%s/%s' % (
                 self['base_url'], self['id'], self['basename_escaped'] )

      if var == 'info_url':
         return '%s/info/%s/%s' % (
                 self['base_url'], self['id'], self['basename_escaped'] )

      if var == 'deletion_url':
         return '%s/delete/%s' % (self['base_url'], self['delete_id'])

      if var == 'basename':
         return basename(self._record['file'])

      if var == 'basename_escaped':
         return urlencode({'': self['basename']})[1:]

      raise KeyError


def yafu_list(db, args):

   offset = args.number - len(db) if args.number else 0

   for record in db:
      offset += 1
      if offset < 0:
         continue

      record = RecordExpander(record, args)

      if args.show_expired or not record['expired']:
         print(args.format.safe_substitute(record))


def yafu_delete(url):

   params = { 'confirm': 'do it!' }
   response = requests.post(url, data=params)

def yafu_upload(file, args):

   ul_files = { 'upload': open(file, 'rb').read() }
   # https://stackoverflow.com/questions/33717690/python-requests-post-with-unicode-filenames

   ul_parameters = {
      'filename':    basename(file),
      'expires':     args.expires,
      'email':       args.email,
      'password':    args.password,
      'comment':     args.comment
   }

   if args.hide:
      ul_parameters['hide'] = 'true'

   response = requests.post(args.base_url + '/index.php', files=ul_files, data=ul_parameters)

   try:
      match = re.search('href="[^"]*/info/([0-9]+)/[^"]+', response.text)
      ul_id = match.group(1)
   except:
      ul_id = 'NOT_AVAILABLE'
      print('Could not extract file id')

   try:
      match = re.search('http://.+/delete/([^"]+)', response.text)
      delete_id = match.group(1)
   except:
      delete_id = 'NOT_AVAILABLE'
      print('Could not extract delete id')

   record = {
      'file':        file,
      'base_url':    args.base_url,
      'id':          ul_id,
      'delete_id':   delete_id,
      'expires':     args.expires,
      'email':       args.email,
      'date_ts':     int( time() ),
      'hide':        args.hide,
      'expire_ts':   0
   }

   if record['expires'] != 'max':
      n = int(record['expires'][:-1])
      t = { 'm':'minutes', 'h':'hours', 'd':'days', 'w':'weeks' }[ record['expires'][-1] ]

      td = timedelta(**{t:n})

      record['expire_ts'] = int( ( datetime.now() + td ).timestamp() )

   return record

args = parser.parse_args()

try:
   record_storage = pickle.load(open(args.db, 'rb'))
except:   
   record_storage = []

if args.upload:
   for file in args.upload:
      record = yafu_upload(file, args)

      fmt_record = RecordExpander(record, args)
      print(args.format.safe_substitute(fmt_record))

      record_storage.append(record)

elif args.delete:
   for url in args.delete:
      yafu_delete(url)

elif args.list:
   yafu_list(record_storage, args) 

pickle.dump(record_storage, open(args.db, 'wb'))

File: yafu.pl
#!/usr/bin/env perl

eval 'exec /usr/bin/perl  -S $0 ${1+"$@"}'
    if 0; # not running under some shell

=begin COPYRIGHT

   yafu.pl - Upload Files to Yet-Another-File-Upload Services
   Copyright (C) 2015 Benjamin Abendroth
   
   This program is free software: you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation, either version 3 of the License, or
   (at your option) any later version.
   
   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.
   
   You should have received a copy of the GNU General Public License
   along with this program.  If not, see <http://www.gnu.org/licenses/>.

=end COPYRIGHT

=cut

use 5.010001;
use feature 'say';

use strict;
use warnings;
use autodie;

use Try::Tiny;
use Tie::File;
use Time::Piece;
use Time::Seconds;
use WWW::Mechanize;
use File::Basename qw(basename);
use URI::Escape qw(uri_escape);
use Config::General qw(ParseConfig);
use Getopt::Long qw(:config gnu_getopt auto_version);

use constant DB_FIELDS => qw(id delete_id base file date expire_date expires hidden);
use constant DB_DELIMITER => "\t";

use constant CONFIG_DIR => $ENV{HOME} . '/.config/yafu';
use constant DEFAULT_CONFIG => CONFIG_DIR . '/yafu.rc';

use constant SHORTCUTS => qw(
   b basename
   f file
   i info_url
   d delete_url
   u direct_url
   D date
   e expire_date
   x expired
   E expires
   h hidden
);

use constant DEFAULT_FORMAT => q(=== %basename% {expired?[EXPIRED] }===
File: %file%
Date: %date%
Expires: %expire_date% [%expires%]
Type: {hidden?Private:Public}
URL: %direct_url%
Info: %info_url%
Delete: %deletion_url%
);

$main::VERSION = "0.6";

# for --list and --delete
my %actions;

# we save our Formatter-Obj here
my $formatter;

# application settings (also available via config file)
my %options = (
   'base'          => 'http://pixelbanane.de/yafu',

   'db'            => CONFIG_DIR . '/yafu.db',

   'email'         => '',
   'password'      => '',
   'comment'       => '',
   'hidden'        => 1,
   'expires'       => '1w',

   'format'        => DEFAULT_FORMAT,
   'show-expired'  => 0,
   'number'        => 0
);

exit main();

sub main
{
   init();

   my @yafudb;
   my $exit = 0;

   if ($options{db} ne '') {
      tie @yafudb, 'Tie::File', $options{db}
         or die "Could not open yafudb: $!";
   }

   if ($actions{delete}) {
      die "Missing arguments for --delete\n" if ! @ARGV;

      for my $link (@ARGV) {
         try
         {
            delete_url($link, \@yafudb);
         }
         catch
         {
            $exit = 1;
            say STDERR "Error while deleting '$link': $_";
         };
      }
   }
   elsif (@ARGV) {
      for my $file (@ARGV) {
         try
         {
            upload($file, \@yafudb);
         }
         catch
         {
            $exit = 1;
            say STDERR "Error while uploading '$file': $_";
         };
      }
   }

   if (! @ARGV || $actions{list}) {
      try
      {
         list_urls(\@yafudb);
      }
      catch
      {
         $exit = 1;
         say STDERR "Error while listing: $_";
      };
   }

   untie @yafudb if ($options{db} ne ''); 
   return $exit;
}

exit main;

sub checkOptions
{
   die "Invalid value for 'expires'\n"
      if $options{expires} !~ /^(30m|1h|6h|1d|3d|1w|max)$/;
}

sub init
{
   # setup config dir
   -d CONFIG_DIR || mkdir CONFIG_DIR || die "Config dir not available: $!\n";

   # process "pseudo-options" and actions first
   Getopt::Long::Configure qw(pass_through);

   my $getopt_config;
   my $getopt_no_config;

   GetOptions(
      'help|h'          => \&print_help,
      'example-config'  => \&print_config,
      'config=s'        => \$getopt_config,
      'no-config'       => \$getopt_no_config,
      'list|l'          => \$actions{list},
      'delete|d'        => \$actions{delete}
   );

   # read the config file
   if (! $getopt_no_config)
   {
      # use --config (or DEFAULT_CONFIG *if* it exists)
      my $rcFile = $getopt_config || -e DEFAULT_CONFIG && DEFAULT_CONFIG;

      if ($rcFile)
      {
         my %rcHash = ParseConfig(
            -ConfigFile => $rcFile,
            -AllowMultiOptions => 'no',
            -InterPolateEnv => 'yes'
         );

         for (keys %rcHash) {
            die "Config file: Unknown option: $_\n" if ! exists $options{$_};
         }

         %options = (%options, %rcHash);

         try {
            checkOptions
         }
         catch {
            die "Config File: $_";
         };
      }
   }

   # now read (and overwrite) the "real options"
   Getopt::Long::Configure qw(no_pass_through);

   GetOptions(\%options,
      'base|b=s',

      'db=s',
      'no-db'           => sub { $options{db} = '' },

      'format|f=s',
      'show-expired|x',
      'number|n=i',

      'email|e=s',
      'password|p=s',
      'comment|c=s',
      'expires|E=s',
      'private'         => sub { $options{hidden} = 1 },
      'public'          => sub { $options{hidden} = 0 }
   ) or exit 1;

   try {
      checkOptions
   }
   catch {
      die "Command Line: $_";
   };
}

sub print_help
{
   require Pod::Usage;
   Pod::Usage::pod2usage(-exitstatus => 0, -verbose => 2);
}

sub print_config
{
   for (keys %options)
   {
      my $value = $options{$_};
      $value =~ s/\n/\\n/smg;

      say "$_ = $value";
   }

   exit 0;
}

sub writeRecord
{
   my ($record, $db) = @_;
   push @$db, join(DB_DELIMITER, @$record{(DB_FIELDS)});
}

sub readRecord
{
   my $line = shift;

   my @fields = split(DB_DELIMITER, $line);

   return map { (DB_FIELDS)[$_] => $fields[$_] } (0..$#fields);
}

sub printRecord
{
   my ($record, $format) = @_;

   $formatter = Formatter->new($format) if ! $formatter;

   my $formatted = $formatter->fill(
         sub { getRecordVar($record, $_[0]) },
         sub { getRecordVar($record, $_[0]) }
   );

   say $formatted if $formatted;
}

sub getRecordVar
{
   my ($rec, $var) = @_;

   if ($var =~ /^\w$/) {
      die "Shortcut $var not found for record\n" if ! {SHORTCUTS}->{$var};
      $var = {SHORTCUTS}->{$var};
   }

   return $rec->{$var} if exists $rec->{$var};

   return $rec->{expired} =
      ($rec->{expires} ne 'max' &&
       $rec->{expire_date} lt localtime->strftime('%Y-%m-%d %H:%M:%S'))
      if $var eq 'expired';

   return $rec->{basename} = basename($rec->{file})
      if $var eq 'basename';

   return $rec->{basename_e} = uri_escape(getRecordVar($rec, 'basename'))
      if $var eq 'basename_e';

   return qq($rec->{base}/$rec->{id}/) . getRecordVar($rec, 'basename_e')
      if $var eq 'direct_url';

   return qq($rec->{base}/info/$rec->{id}/) . getRecordVar($rec, 'basename_e')
      if $var eq 'info_url';

   return qq($rec->{base}/delete/$rec->{delete_id})
      if $var eq 'deletion_url';

   die "Key $var not found in record\n";
}

sub upload
{
   my ($file, $db) = @_;

   my %params = %options{'email', 'expires', 'password', 'comment'};
   $params{upload} = $file;
   $params{filename} = basename($file);
   $params{hidden} = 'true' if $options{hidden};

   my $mech = WWW::Mechanize->new;
   $mech->get($options{base} . '/index.php');
   $mech->submit_form(
         form_id  => 'UploadForm',
         fields => \%params
   );

   my %record = %options{'base', 'hidden', 'expires'};
   $record{file} = $file;

   unless ( ($record{id}) =
         $mech->content =~ m!href="[^"]*/info/([0-9]+)/[^"]+!)
   {
      $record{id} = 'n/a';
   }

   unless ( ($record{delete_id}) =
         $mech->content =~ m!http://.+/delete/([^"]+)!)
   {
      $record{delete_id} = 'n/a';
   }

   if ($record{id} eq 'n/a' || $record{delete_id} eq 'n/a')
   {
      say STDERR 'Could not extract all information';

      try      
      {
         require File::Temp;

         my ($temp_fh, $temp_name) = File::Temp::tempfile();
         print $temp_fh $mech->content; 
         say STDERR "Dumped response to '$temp_name'";
         close $temp_fh;
      }
      catch
      {
         say STDERR "Could not dump respone: $_";
      };
   }

   my $now = localtime;
   $record{date} = $now->strftime('%Y-%m-%d %H:%M:%S');

   if (my ($repeat, $type) = $record{expires} =~ /([0-9]+)([a-z])/)
   {
      $record{expire_date} =
      (
         $now + $repeat *
         {
            m => ONE_MINUTE,
            h => ONE_HOUR,
            d => ONE_DAY,
            w => ONE_WEEK

         }->{$type}

      )->strftime('%Y-%m-%d %H:%M:%S');
   }
   else
   {
      $record{expire_date} = 0;
   }

   writeRecord(\%record, $db);
   printRecord(\%record, $options{format});
}

sub delete_url
{
   my ($url, $db) = @_;

   my $mech = WWW::Mechanize->new;
   $mech->get($url);
   $mech->submit_form( button => 'confirm');
}

sub list_urls
{
   my $db = shift;

   my $start = 0;

   if ($options{number} && $options{number} < scalar @$db)
   {
      $start = scalar @$db - $options{number};
   }

   for my $i ($start..$#{ $db })
   {
      my %record = readRecord($db->[$i]);

      next if (! $options{'show-expired'} && getRecordVar(\%record, 'expired'));

      printRecord(\%record, $options{format});
   }
}

package Formatter::List
{
   sub new
   {
      return bless [], shift;
   }

   sub add
   {
      my ($self, $child) = @_;

      # if $child is charater and prev. element is string
      if ( !ref($child) && @$self && !ref($self->[-1]) )
      {
         # append $child to string
         $self->[-1] .= $child;
      }
      else
      {
         push @$self, $child;
      }
   }

   sub fill
   {
      my ($self, $condCallback, $valueCallback) = @_;

      my $result = '';

      for (@$self)
      {
         if (! ref $_) {
            $result .= $_;
         }
         else {
            $result .= $_->fill($condCallback, $valueCallback);
         }
      }

      return $result;
   }
}

package Formatter::Variable
{
   sub new
   {
      my ($class, $var) = @_;
      return bless \$var, $class;
   }

   sub fill
   {
      my ($self, $condCallback, $valueCallback) = @_;
      return $valueCallback->($$self);
   }
}

package Formatter::Conditional
{
   sub new
   {
      my ($class, $var) = @_;

      return bless {
         variable => $var,
         state => 'true',
         true => Formatter::List->new,
         false => Formatter::List->new
      }, $class;
   }

   sub add
   {
      my ($self, $child) = @_;

      $self->{ $self->{state} }->add($child);
   }

   sub flip
   {
      $_[0]->{state} = 'false';
   }

   sub fill
   {
      my ($self, $condCallback, $valueCallback) = @_;

      if ( $condCallback->( $self->{variable} ) ) {
         return $self->{true}->fill($condCallback, $valueCallback);
      }
      else {
         return $self->{false}->fill($condCallback, $valueCallback);
      }
   }
}

package Formatter
{
   sub new
   {
      my ($class, $format) = @_;

      open(my $fmt_handle, '<', \$format);
      local *STDIN = *$fmt_handle;

      return Formatter::parse( Formatter::List->new );
   }

   sub parse
   {
      my $obj = shift;

      while (my $c = getc)
      {
         if ($c eq '%')
         {
            my $var = '';
            $var .= $c while ( ($c = getc) && $c ne '%' );

            die "Missing variable name after '%'" unless $var;
            die "Trailing '%' not found" unless $c;

            $obj->add(Formatter::Variable->new($var));
         }
         elsif ($c eq '{')
         {
            my $var = '';
            $var .= $c while ( ($c = getc) && $c ne '?' );

            die "Missing variable name after '{'" unless $var;
            die "Missing '?' inside '{...}'" unless $c;

            my $n_obj = Formatter::Conditional->new($var);

            Formatter::parse($n_obj);

            $obj->add($n_obj);
         }
         elsif ($c eq '}')
         {
            die "Closing '}' without beginning '{'" if ! $obj->isa('Formatter::Conditional');
            return $obj;
         }
         elsif ($c eq ':' && $obj->isa('Formatter::Conditional') && $obj->{state} eq 'true')
         {
            $obj->flip;
         }
         else
         {
            if ($c eq '\\')
            {
               $c = getc || die "Missing character after '\\'";

               next if ($c eq "\n");

               $c = {
                  n => "\n",
                  t => "\t" 
               }->{$c} || $c;
            }

            $obj->add($c);
         }
      }

      die "Missing closing '}'" if $obj->isa("Formatter::Conditional");
      return $obj;
   }
}

__END__

=pod 

=head1 NAME

yafu.pl - Upload Files to Yet-Another-File-Upload Services

=head1 SYNOPSIS

=over

yafu.pl [I<OPTION>]... [ARGS]...

=back

=head1 OPTIONS

=head2 Basic Startup Options

=over

=item B<--help>

Display this help text and exit.

=item B<--version>

Display the script version and exit.

=item B<--example-config>

Display an example configuration file and exit.

=back

=head2 Actions

=over

=item B<--delete|-d>

Delete the URLs given on the rest of the command line.

=item B<--list|-l>

List all files including information.

=back

=head2 General Options

=over

=item B<--config> I<file>

Use config file given in I<file>.

=item B<--no-config>

Don't use a configuration file at all.

=item B<--db> I<file>

Use I<file> as database file.

=item B<--no-db>

Don't use a database file at all.

=back

=head2 List Options

=over

=item B<--show-expired>

Also list expired files.

=item B<--number|-n> I<NUM>

Only print out the last I<NUM> records.

=item B<--format|-f> I<format>

Specify output format.

Variables are enclosed inside I<%...%>.

Conditional expressions: B<{>I<VARIABLE>B<?>I<print if true>B<:>I<print if result>B<}>.

Escaping I<char>s like "B<{>", "B<%>" and "B<:>" via B<\>I<char>.

Available variables are:
B<file|f>, B<basename|b>,
B<direct_url|u>, B<deletion_url|d>, B<info_url|i>
B<date|D>, B<expire_date|e>, B<expires|E>,
B<id>, B<delete_id>

Boolean variables:
B<expired|x>, B<hidden|h> 

B<Example>:

"URL: %I<direct_url>% {I<expired>?EXPIRED:%I<deletion_url>%}"

Will print the I<direct_url>. It will also print the I<deletion_url> if I<expired> is B<false>, else it will print "EXPIRED".

=back

=head2 Upload Options

=over

=item B<--base> I<URL>

Use I<URL> as base url for yafu. Defaults to I<http://pixelbanane.de/yafu>.

=item B<--email> I<email>

Use I<email> for email field.

=item B<--password> I<password>

Use I<password> as password.

=item B<--comment> I<comment>

Use I<comment> as comment.

=item B<--private>

Don't put the file in the public file list. This is the default.

=item B<--public>

Put the file in the public file list.

=item B<--expires> I<time>

Set when the file will expire on the server. I<time> can be:

B<30m>: 30 minutes,
B<1h>: 1 hour, B<6h>: 6 hours,
B<1d>: 1 day, B<3d>: 3 days,
B<1w>: 1 week,
B<max>: Till the very end

Defaults to I<1w>.

=back   

=head1 CONFIGURATION

=over

See B<--example-config>.

Environment variables like B<$HOME> can be used insied the configuration file.

Example:

B<expires> = "I<1d>"

B<comment> = "I<my standard comment>"

B<hidden> = I<0>

B<db> = I<$HOME/yafu.db>

=back

=head1 FILES

=over

=item I<$HOME/.config/yafu/yafu.rc>

Default configuration file. See B<--config>.

=item I<$HOME/.config/yafu/yafu.db>

Default database file. See B<--db>.

=back

=head1 AUTHOR

Written by Benjamin Abendroth.

=cut

